### PR TITLE
docs: clarify subdomain CID normalization

### PR DESCRIPTION
**Pull request by lidel**
This PR adds more detail into "best practices" when implementing `ipfs://` URI backed by a subdomain gateway.

Namely, opening a path with CID as-is, to take advantage of DNS normalization built into go-ipfs 0.5+

This should help with native URIs in Brave (https://github.com/brave/brave-browser/issues/10220) when embedded go-ipfs is turned off and other collabs.

### CODE REVIEW

File: ADDRESSING.md [hunk 1]
```diff
@@ -20,18 +20,20 @@
...
-If no native protocol handler is available, redirect to a gateway. The implementation should detect if a local IPFS node is available. In web browser contexts where a local IPFS node is present, use [subdomain gateway](https://docs.ipfs.io/how-to/address-ipfs-on-web/#subdomain-gateway) at `localhost`. If not, use public one such as `dweb.link`:
+If no native protocol handler is available, redirect to IPFS path at a gateway. The implementation should detect if a local IPFS node is available. In web browser contexts where a local IPFS node is present, use [subdomain gateway](https://docs.ipfs.io/how-to/address-ipfs-on-web/#subdomain-gateway) at `localhost`. If not, use public one such as `dweb.link`. 
+
+In both cases point at IPFS path first, to ensure gateway takes care of CID normalization into a DNS-safe form:
 
 ```bash
-ipfs://{cid}                → https://{cid}.ipfs.dweb.link
-ipns://{libp2p-key}         → https://{libp2p-key}.ipns.dweb.link
+ipfs://{cid}          → https://dweb.link/ipfs/{cid}        → HTTP301 → https://{dns-safe-cid}.ipfs.dweb.link
+ipns://{libp2p-key}   → https://dweb.link/ipns/{libp2p-key} → HTTP301 → https://{dns-safe-cid}.ipns.dweb.link
```
Comment (autonome) [inline ADDRESSING.md:29]: I like the clarity of the "HTTP301" but is not clear who is taking which action. I'd prefer that this somehow identify where the boundaries of the implementor's actions are, and where the gateway is.

Also, not required for now, but in the first arrows perhaps we should have a clarifying action like "expand to", like you did "HTTP301" to be clear about what's happening there.

File: ADDRESSING.md [hunk 2]
```diff
@@ -20,18 +20,20 @@
...
 ```
 
 When DNSLink hostname is used, redirect to subdomain only with `localhost`, use original HTTP URL otherwise:
 
 ```
-ipns://{fqdn-with-dnslink}  → https://{fqdn-with-dnslink}.ipns.localhost
-                            → https://{fqdn-with-dnslink} (if no localhost node)
+ipns://{fqdn}         → https://localhost/ipns/{fqdn}       → HTTP301 → https://{fqdn}.ipns.localhost
+                      → https://{fqdn} (if no local node)
```
Comment (autonome) [inline ADDRESSING.md:36]: ditto about the roles/boundaries of implementor vs gateway.

File: ADDRESSING.md [hunk 3]
```diff
@@ -53,13 +55,20 @@ ipfs://{fqdn-with-dnslink} → redirect → ipns://{fqdn-with-dnslink} # just to
 
 When [origin-based security](https://en.wikipedia.org/wiki/Same-origin_policy) perimeter is needed, [CIDv1](https://github.com/ipld/cid#cidv1) in Base32 ([RFC4648](https://tools.ietf.org/html/rfc4648#section-6), no padding) should be used in subdomain:
 
-    https://<cidv1-base32>.ipfs.<gateway-host>.tld/path/to/resource
-    https://<cidv1-base32-libp2p-key>.ipns.<gateway-host>.tld/path/to/resource
+    https://<cid>.ipfs.<gateway-host>.tld/path/to/resource
+    https://<libp2p-key-cid>.ipns.<gateway-host>.tld/path/to/resource
```
Comment (autonome) [inline ADDRESSING.md:60]: Should ensure this notation is consistent throughout the spec.

File: ADDRESSING.md [hunk 4]
```diff
@@ -20,18 +20,20 @@
 
 ## TL;DR
 
-If no native protocol handler is available, redirect to a gateway. The implementation should detect if a local IPFS node is available. In web browser contexts where a local IPFS node is present, use [subdomain gateway](https://docs.ipfs.io/how-to/address-ipfs-on-web/#subdomain-gateway) at `localhost`. If not, use public one such as `dweb.link`:
+If no native protocol handler is available, redirect to IPFS path at a gateway. The implementation should detect if a local IPFS node is available. In web browser contexts where a local IPFS node is present, use [subdomain gateway](https://docs.ipfs.io/how-to/address-ipfs-on-web/#subdomain-gateway) at `localhost`. If not, use public one such as `dweb.link`. 
+
+In both cases point at IPFS path first, to ensure gateway takes care of CID normalization into a DNS-safe form:
```
Comment (johnnymatthews) [inline ADDRESSING.md:25]: > In both cases point at IPFS path first

Is this an instruction, or are we telling the reader that we are pointing to the IPFS path before we do anything else?

File: ADDRESSING.md [hunk 5]
```diff
@@ -53,13 +55,20 @@ ipfs://{fqdn-with-dnslink} → redirect → ipns://{fqdn-with-dnslink} # just to
...
-    https://<cidv1-base32>.ipfs.<gateway-host>.tld/path/to/resource
-    https://<cidv1-base32-libp2p-key>.ipns.<gateway-host>.tld/path/to/resource
+    https://<cid>.ipfs.<gateway-host>.tld/path/to/resource
+    https://<libp2p-key-cid>.ipns.<gateway-host>.tld/path/to/resource
 
 Example:
 
     https://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.dweb.link/wiki/
 
+Subdomain gateway feature in go-ipfs takes care of necessary CID conversion when IPFS path is requested:
```
Comment (johnnymatthews) [inline ADDRESSING.md:65]:
```suggestion
The subdomain gateway feature in go-ipfs takes care of necessary CID conversion when IPFS path is requested:
```

**Comment by johnnymatthews** [comment]
Added a couple of notes. I'm not too clear on the purpose of this doc. Is this a guide instructing the reader to do something? Or are we saying _these things happen in this feature_?

File: ADDRESSING.md (hunk 4 above)
Comment (lidel) [reply ADDRESSING.md:25]: It is instruction / "best practice".

File: ADDRESSING.md [hunk 6]
```diff
@@ -20,18 +20,20 @@
 
 ## TL;DR
 
-If no native protocol handler is available, redirect to a gateway. The implementation should detect if a local IPFS node is available. In web browser contexts where a local IPFS node is present, use [subdomain gateway](https://docs.ipfs.io/how-to/address-ipfs-on-web/#subdomain-gateway) at `localhost`. If not, use public one such as `dweb.link`:
+If no native protocol handler is available, redirect to IPFS path at a gateway. The implementation should detect if a local IPFS node is available. In web browser contexts where a local IPFS node is present, use [subdomain gateway](https://docs.ipfs.io/how-to/address-ipfs-on-web/#subdomain-gateway) at `localhost`. If not, use public one such as `dweb.link`.
```
[the remark this answers is not in the document]
Comment (lidel) [reply ADDRESSING.md]: Fixed in [`9b3375a` (#164)](https://github.com/ipfs/in-web-browsers/pull/164/commits/9b3375a6fdefe399a255369f62a4385147d0cb1d)

File: ADDRESSING.md (hunk 1 above)
Comment (lidel) [reply ADDRESSING.md:31]: I've added abstract header in [`21a4fad` (#164)](https://github.com/ipfs/in-web-browsers/pull/164/commits/21a4fadbbc17cf407ab812d8eb1592a37f439a41) that clarifies the boundary a bit, but did not add "expand to", signal/noise ratio did not felt be beneficial:

```
Native URI            – HTTP Gateway                       (– Internal normalization done by HTTP Gateway)   
ipfs://{cid}          → https://dweb.link/ipfs/{cid}        → HTTP301 → https://{dns-safe-cid}.ipfs.dweb.link
```

**Comment by lidel** [comment]
I've applied suggestions from reviews.

@johnnymatthews  – mind taking a final look?

File: ADDRESSING.md (hunk 4 above)
Comment (johnnymatthews) [reply ADDRESSING.md:25]:
```suggestion
In either case, point at IPFS path first to ensure gateway takes care of CID normalization into a DNS-safe form:
```

**Comment by johnnymatthews** [comment]
The rest looks solid to me! Just that one tiny change 😄
